### PR TITLE
Quote Multipart/Form-Data Field Names

### LINKED ISSUE #6780: Invoke-WebRequest multipart form-data fails file upload when field names not quoted
This happens using the new `-Form` parameter that takes a hashtable of values and creates the multi-part form data.

Steps to reproduce
------------------
1. Find URL to webservice that accepts POST of multi-part form data
2. Run Fiddler or Wireshark to observe the request body
3. Execute IWR similar to below:

```powershell
Invoke-WebRequest -Uri $Uri -Method POST -Form @{
    APPLICATION_NAME = "Name"
    APPLICATION_VERION = "3.0"
    UPLOADED_FILE = Get-Item .\foo.zip
}
```

Expected behavior
-----------------
Upload works as it does when using Python script to upload the file.  Observe that in this Fiddler trace, that the form-data field values are not quoted:

image

Actual behavior
---------------
PowerShell fails and after some experimenting it is because the name values are not quoted:

image

With just a few extra quotes e.g.:

```powershell
Invoke-WebRequest -Uri $Uri -Method POST -Form @{
    '"APPLICATION_NAME"' = "Name"
    '"APPLICATION_VERION"' = "3.0"
    '"UPLOADED_FILE"' = Get-Item .\foo.zip
}
```

I can get the PowerShell request to quote all the field data values **except** the file's `filename` field.  And that is enough to break the upload.  Besides that, the above workaround is ugly.

As an aside, is there a way to control the filename value?  If there was, I could work-around this by supplying the quotes myself.  Also, I could see someone wanting to change the filename to something other than the original file's filename.

This also makes me wonder if we could use a new parameter like `-RequestFilter` that took a scriptblock with a signature like `{param($requestText) ... }` where you could get the request text just before it is sent.  Then you could manipulate it however you needed to and return that text from this scriptblock.  With such a feature I might be able to do this:

```powershell
Invoke-WebRequest -Uri $Uri -Method POST `
    -RequesetFilter {param($r) $r -replace '(?<=filename=)(\S+)','"$1"' } `
    -Form @{
        '"APPLICATION_NAME"' = "Name"
        '"APPLICATION_VERION"' = "3.0"
        '"UPLOADED_FILE"' = Get-Item .\foo.zip
    }
```

Environment data
----------------

```powershell
> $PSVersionTable
Name                           Value
----                           -----
PSVersion                      6.1.0-preview.2
PSEdition                      Core
GitCommitId                    v6.1.0-preview.2
OS                             Microsoft Windows 10.0.16299
Platform                       Win32NT
PSCompatibleVersions           {1.0, 2.0, 3.0, 4.0...}
PSRemotingProtocolVersion      2.3
SerializationVersion           1.1.0.1
WSManStackVersion              3.0
```

### PR DESCRIPTION
**Pull request by markekraus**
## PR Summary

* fixes  #6780
* When `-Form` is used, the field names and the filename value will be enclosed in quotes
* fixes multipart/form-data submission issues with some APIs
* brings functionality closer to modern browsers and curl

Since `-Form` has not been released in an actual release version, this is not a breaking change. 

There is no decent way to test this as the metadata for text fields is lost when the form is processed in ASP.NET.  Only way would be to return the raw request body and parse it with regex.

## PR Checklist

- [X] [PR has a meaningful title](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
    - Use the present tense and imperative mood when describing your changes
- [X] [Summarized changes](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [X] [Change is not breaking](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#making-breaking-changes)
- [X] [Make sure all `.h`, `.cpp`, `.cs`, `.ps1` and `.psm1` files have the correct copyright header](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [X] This PR is ready to merge and is not [Work in Progress](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---work-in-progress).
    - If the PR is work in progress, please add the prefix `WIP:` to the beginning of the title and remove the prefix when the PR is ready.
- **User-facing changes**
    - [X] Not Applicable
    - **OR**
    - [ ] User-facing [Documentation needed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
        - [ ] Issue filed - Issue link:
- **Testing - New and feature**
    - [X] Not Applicable or can only be tested interactively
    - **OR**
    - [ ] [Make sure you've added a new test if existing tests do not effectively test the code changed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#before-submitting)
        - [X] [Add `[feature]` if the change is significant or affects feature tests](https://github.com/PowerShell/PowerShell/blob/master/docs/testing-guidelines/testing-guidelines.md#requesting-additional-tests-for-a-pr)

### CODE REVIEW

**Comment by SteveL-MSFT** [comment]
Should we open an issue against corefx?  If corefx supports/fixes this in the future, will we have double quotes?

**Comment by markekraus** [comment]
At this point it would be a breaking change in CoreFX. This is the same behavior for `System.Net.Http.MultipartFormDataContent` in .NET Framework 4.5 when it was added. I found a [blog article form 2013](https://blog.xdumaine.com/uploading-a-photo-as-multipart-form-data-in-c/) that references this issue.  If they did add this, they would probably do so through a new property in which case we wouldn't be broken.

**Comment by markekraus** [comment]
I found this https://github.com/dotnet/corefx/issues/26886#issuecomment-381697212 though, this begs the question why every browser and curl wraps the name and filename fields values in quotes and .NET does not?

**Review by SteveL-MSFT — APPROVED** [review]
LGTM

**Comment by iSazonov** [comment]
Seems it is important fix so we need to add tests. Maybe use output with `-WhatIf`?

**Comment by iSazonov** [comment]
I believe we should have tests for the uploading.
I assumed that we could do verbose output of the form but seems it is impossible. So we need full feature tests. Perhaps @JamesWTruher and @adityapatwardhan have ideas about this tests.

**Comment by SteveL-MSFT** [comment]
@markekraus I think it would be ok to add a test to weblistener to specifically return the field names and values as raw text.  Since the values are controlled by the Pester test, it should know exactly what should be returned without having to resort to regex

**Comment by markekraus** [comment]
@SteveL-MSFT The problem is the ASP.NET does not provide whether or not the feild names are quoted when you consume the response as a form. The only way to test that is to echo back the raw HTTP request as a string.

**Comment by SteveL-MSFT** [comment]
@markekraus I see, so you either get just the name value pairs stripped of the quotes or the entire raw request.  It seems the risk is low and the cost is high.  I would be fine creating a new issue to add such a test in the future, particularly if it's useful beyond this specific issue.

**Comment by markekraus** [comment]
Added #6794 to track possible adding an echo controller to WebListener.

**Comment by iSazonov** [comment]
>the cost is high

@markekraus implemented WebListener  so fast that I'm still amazed :-)

**Review by iSazonov — DISMISSED** [review]

**Comment by iSazonov** [comment]
@daxian-dbw Do you agree that we use the `$""` string interpolation instead of explicit formatting?

**Comment by daxian-dbw** [comment]
I don't have a strong opinion. I don't want to block people from using C# language features :)
When converting an interpolated string to a string instance, the current culture is used, which seems fine in this change. But be careful if you want a culture-invariant string.

**Comment by iSazonov** [comment]
We could
```c#
 = "\"" + ... + "\"";
```

**Comment by iSazonov** [comment]
Sorry for wrong close.

**Comment by daxian-dbw** [comment]
You can also choose culture with the string interpolation:
```c#
FormattableString message = $"The speed of light is {speedOfLight:N3} km/s.";
string messageInInvariantCulture = FormattableString.Invariant(message);
```
See https://docs.microsoft.com/en-us/dotnet/csharp/language-reference/tokens/interpolated

**Comment by iSazonov** [comment]
I think that the double formatting superfluous here. 
Concatenating enough.
Or we could use `String.Create()` from .Net Core 2.1 (later).

**Comment by markekraus** [comment]
@iSazonov Does the culture on these even matter? I'm not doing any format transformations, I'm just shoving one string inside another.

**Comment by iSazonov** [comment]
I remember that we have already corrected misuses of culture in formatting. So I'm trying to be careful here. 
If culture is not important here, we should not use this pattern at all.
Now we have already merged #6718 and can use .Net Core 2.1 features - it is very optimized for string manipulations. 
I believe here we could use `String.Create()`. Samples: 
https://msdn.microsoft.com/en-us/magazine/mt814808.aspx
http://source.dot.net/#System.Private.Uri/System/Uri.cs,1412

**Comment by iSazonov** [comment]
>So I'm asking.. does it even matter here?

No. The discussion is about _code style_. Bad style is a open door for camouflaged bugs. Therefore, we require explicit curly braces in the `if` and explicit culture in formattings - after fixing and discussing the latter we even added manuals to our documentation #4983 
Here I wanted to understand whether we can use `$""` and now believe that we should not use this.

**Comment by markekraus** [comment]
... why is this bad style? what is wrong with `$""`?  I really don't understand what your objecttion is. You don't like it?

**Comment by markekraus** [comment]
@iSazonov Also, I cannot figure out how you intend to use String.Create() here. Can you provide me with an example relevant to this PR? It seems like it is overkill for string concatenation.

**Comment by markekraus** [comment]
ok.. so `$""` is banned even when used properly? So what is the best way to concat strings in this project then? using `+` has a perf hit so that's out. `String.Concat()` is in the same boat and needs multiple calls. `String.Create()` is confusing (seriously.. what is a `TState`? why is that not documented anywhere? how exactly would you even use this to concat string?) In this case, `String.Format()` and `$""` are identical. I could use `String.Format()` with the explicit formatter, but that seems really verbose for just enclosing a string in quotes.

**Comment by iSazonov** [comment]
>ok.. so $"" is banned even when used properly?

If other maintainers agree that we should always use _explicit_ culture then yes.

Previously we use + for concatinations for few arguments (~< 5). Othervise StringBuilder is preferable.
Currently we can and should use .Net Core 2.1 and C# 7.2 features to decrease allocations.
Now we can use `ValueStringBuilder`, `String.Create()`, Span<T> and many new string method overloads.

In the case I'd use `String.Create()`. Although you can ignore it now and use +.
Sample:
```c#
            var name = LanguagePrimitives.ConvertTo<String>(fieldName);
            contentDisposition.Name = String.Create(2+name.Length, name, (Span<char> destination, string str) =>
            {
                destination[0] = '"';
                str.AsSpan().CopyTo(destination.Slice(1));
                destination[destination.Length-1] = '"';
            });
```

**Comment by markekraus** [comment]
> If other maintainers agree that we should always use explicit culture then yes.

Why? Why always use explicit culture? String concatenation does not require it. It _is_ relvant when doing formatting or transformations, but just combining strings it's senseless.

> Previously we use + for concatinations for few arguments (~< 5). 

If this will unblock this PR so it will be merged, I will change it to use `+`. I still challenge the logic behind this debate.

> In the case I'd use String.Create(). 

Why? That is a seriously over-complicated verbose and tough to follow piece code just to put `"` around a string. I get it if we were working with concatenating some large strings or this code was being called so many times that performance really matters. but this is a fairly esoteric feature that wont be used often or heavily and it's more important the the code be readable here than to squeeze every ounce of performance out of it.

**Comment by iSazonov** [comment]
> this is a fairly esoteric feature that wont be used often or heavily

We never know exactly what users will do.

**Comment by markekraus** [comment]
> We never know exactly what users will do.

We do know that this is a form submission style that is not as widely adopted as standard forms. We do know that the intended purpose of this submission method is GUI browser based communications. We do know that some APIs do use this for file uploads, but not all. We do know that while there have been asks for this feature to be added, not many. 

If I didn't have my own personal need for it, it may never have even entered consideration to be adopted in the web cmdlets.

Yes, we are not clairvoyant and we cannot know 100% what users will do, but, let's be realistic here: the string concatenation for quoting field names filename values and of this feature is NEVER going to be a performance concern. Especially when it is already associated with slow, long running, and memory intense tasks (uploading data). The few fractions of a millisecond shaved off and memory usage saved by using `String.Create()` here is severely outweighed by a multitude of other critical performance losses happening at the same time in the same feature.

What I do know for certain is that I will have to look at this code in the future. And very likely, someone else will too. I do know that String.Create() is complex and not easy to read. I do know that I can skim the `$""` method and see very quickly what is being done. I do know that It takes me a few seconds to figure out that the code you provided is just wrapping a string in quotes.

If you want to add a string concatenation utility method that allows us to abstract `String.Create()` throughout the project, I would consider using that. But I don't see the need to add such a thing directly in the web cmdlets. I also don't see any benefit at all in using direct `String.Create()` calls here either.

**Comment by iSazonov** [comment]
Forget about `String.Create()` - you could use `String.Concat()` or +.

>I do know that I can skim the $"" method and see very quickly what is being done. I do know that It takes me a few seconds to figure out that the code you provided is just wrapping a string in quotes.

This is not obviously for code reviewer. My first impression was that it evidently. Then there were questions. Why we use formatting if we need quoting? Why we use `InvariantCulture` in `ConvertTo` and `CurrentCulture` in `$""`? What's the trick? Should we worry about allocations? - we just had a discussion that PowerShell sometimes consumes too much resources and we should be more thrifty. As a result, we received a good exchange of views. It was useful for me to know your point of view.

Summing up this discussion I think `String.Concat()` will be the best pattern here.

**Comment by rkeithhill** [comment]
In this case, the compiler will translate the `+` concatenated strings to a call to String.Concat:

image

**Comment by iSazonov** [comment]
@rkeithhill What plugin you use in VS Code to see IL? In what will be compiled several operators `"a"+"b"+"c"`?

**Comment by rkeithhill** [comment]
I use JetBrains' dotPeek.  Since `String.Concat()` has an overload that takes `params string[]`, I assume it take as many as you supply it.  But I haven't tested that. Here it is with 8 concats:
image

**Comment by rkeithhill** [comment]
It would be much appreciated if this PR makes it in for Preview3.  Just sayin'.  :-)
